File: Blind 75/06 - Linked List/143-reorder-list.py
'''
You are given the head of a singly linked-list. The list can be represented as:

L0 → L1 → … → Ln - 1 → Ln
Reorder the list to be on the following form:

L0 → Ln → L1 → Ln - 1 → L2 → Ln - 2 → …
You may not modify the values in the list's nodes. Only nodes themselves may be changed.
'''
from copy import deepcopy
from time import time
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = head
        fast = head.next
        # Get end of lists
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Reverse second list
        l2 = slow.next
        slow.next = prev = None
        while l2:
            temp = l2.next
            l2.next = prev
            prev = l2
            l2 = temp

        # Assign
        l1, l2 = head, prev
        while l2:
            temp1, temp2 = l1.next, l2.next
            l1.next = l2
            l2.next = temp1
            l1, l2 = temp1, temp2

        return head

    def reference(self, head: Optional[ListNode]) -> None:
        # Find middle
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Reverse second half
        second = slow.next
        prev = slow.next = None
        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp

        # Merge two halfs
        first, second = head, prev
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2

        return head

    def quantify(self, test_cases, runs=10000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                # Create deep copy
                copy = deepcopy(case)
                if i == 0:
                    print(self.reorderList(copy).__dict__)
                else:
                    self.reorderList(copy)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                # Create deep copy
                copy = deepcopy(case)
                if i == 0:
                    print(self.reference(copy).__dict__)
                else:
                    self.reference(copy)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        ListNode(1, ListNode(2, ListNode(3, ListNode(4)))),
        ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5))))),
    ]
    test.quantify(test_cases)

File: Blind 75/06 - Linked List/JavaScript/143-reorder-list.js
const { performance } = require("perf_hooks");

class ListNode {
  constructor(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

class Solution {
  method(head) {
    let [slow, fast] = [head, head.next];

    while (fast && fast.next) {
      slow = slow.next;
      fast = fast.next.next;
    }

    // Reverse p2
    let l2 = slow.next;
    let [prev, temp] = [null, null];
    slow.next = null;
    while (l2) {
      temp = l2.next;
      l2.next = prev;
      prev = l2;
      l2 = temp;
    }

    let l1 = head;
    let [temp1, temp2] = [null, null];
    l2 = prev;
    while (l2) {
      [temp1, temp2] = [l1.next, l2.next];
      l1.next = l2;
      l2.next = temp1;
      [l1, l2] = [temp1, temp2];
    }

    return head;
  }

  reference(head) {
    const mid = this.getMid(head); /* Time O(N) */
    const reveredFromMid = this.reverse(mid); /* Time O(N) */

    this.reorder(head, reveredFromMid); /* Time O(N) */

    return head;
  }

  getMid = (head) => {
    let [slow, fast] = [head, head];

    while (fast && fast.next) {
      /* Time O(N) */
      slow = slow.next;
      fast = fast.next.next;
    }

    return slow;
  };

  reverse = (head) => {
    let [prev, curr, next] = [null, head, null];

    while (curr) {
      /* Time O(N) */
      next = curr.next;
      curr.next = prev;

      prev = curr;
      curr = next;
    }

    return prev;
  };

  reorder = (l1, l2) => {
    let [first, next, second] = [l1, null, l2];

    while (second.next) {
      /* Time O(N) */
      next = first.next;
      first.next = second;
      first = next;

      next = second.next;
      second.next = first;
      second = next;
    }
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        // Create deep copy
        input = JSON.parse(JSON.stringify(input));
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        // Create deep copy
        input = JSON.parse(JSON.stringify(input));
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4)))),
  new ListNode(
    1,
    new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5))))
  ),
];
test.quantify(testCases);
